app/presenters/tree_builder_vms_and_templates.rb

```diff
@@ -8,18 +8,18 @@ def initialize(root, *args)
 
   def relationship_tree
     # TODO: Do more to pre-prune the tree.
-    # - Use :of_type to limit the types of objects
     # - Perhaps only get the folders, then prune those based on RBAC and let
     #   the UI still do the lazy loading of individual folders.  This can be
     #   possibly done by querying the relationship tree only, and then only
     #   converting the folders to real objects.  For the VMs we only need ids,
     #   so taking them from the relationship records can cut down on the huge
     #   VM query.
-    tree = root.subtree_arranged
+    display_vms = Settings.ui.tree.display_vms
+    tree = root.subtree_arranged(display_vms ? {} : {:except_type => VmOrTemplate})
 
     prune_non_vandt_folders(tree)
     reparent_hidden_folders(tree)
-    prune_rbac(tree)
+    prune_rbac(tree) if display_vms
```

Comment:
If the vms are not loaded into memory, then pruning will prune all.

Yes, this is the reason why we are currently avoiding this PR and I made it a WIP.

Anything that gives a performance boost and does not change the behavior has been pulled out into other PRs.